src/Simulation/Common/IQuantumProcessor.cs

```diff
@@ -0,0 +1,640 @@
+        /// <param name="qubit">Qubit to which the gate should be applied.</param>
+        void Reset(Qubit qubit);
+
+        /// <summary>
+        /// Intended for a limited support of branching upon measurement results on a target machine level.
+        /// Called when a conditional statement on measurement results is invoked.
+        /// </summary>
+        /// <param name="measurementResults">The actual results of the measurements of a number of qubits upon which branching is to be performed.</param>
+        /// <param name="resultsValues">The expected values of results of the measurements of these qubits.</param>
+        /// <returns> A value representing this conditional statement and encoding the result of condition.</returns>
+        /// <remarks>
+        /// A typical implementation will compare all <paramref name="measurementResults"/> to <paramref name="resultsValues"/> and return the result of this comparison.
+        /// </remarks>
+        int StartConditionalStatement(IQArray<Result> measurementResults, IQArray<Result> resultsValues);
+
+        /// <summary>
+        /// Intended for a limited support of branching upon measurement results on a target machine level.
+        /// Called when a conditional statement on a measurement result is invoked.
+        /// </summary>
+        /// <param name="measurementResult">The actual result of the measurement of a qubit upon which branching is to be performed.</param>
+        /// <param name="resultValue">The expected value of result of the measurement of this qubit.</param>
+        /// <returns> A value representing this conditional statement and encoding the result of condition.</returns>
+        /// <remarks>
+        /// A typical implementation will compare <paramref name="measurementResult"/> to <paramref name="resultValue"/> and return the result of this comparison.
+        /// </remarks>
+        int StartConditionalStatement(Result measurementResult, Result resultValue);
+
+        /// <summary>
+        /// Intended for a limited support of branching upon measurement results on a target machine level.
+        /// Called when the "then" statement of a conditional statement is about to be executed.
+        /// </summary>
+        /// <param name="statement">A value representing this conditional statement and encoding the result of condition.</param>
+        /// <returns> If true is returned, the "then" statement will be executed, otherwise it will be skipped and RepeatThenClause will not be called.</returns>
+        /// <remarks>
+        /// A typical implementation will use <paramref name="statement"/> to return whether condition was evaluated to true.
+        /// </remarks>
+        bool RunThenClause(int statement);
+
+        /// <summary>
+        /// Intended for a limited support of branching upon measurement results on a target machine level.
+        /// Called when the "then" statement of a conditional statement has finished executing.
```

Comment:
This is clearly explained in the comment to the RunThenClause. It says RepeatThenClause will not be called if RunThenClause returns false. RepeatThenClause also clearly states it will be called only after "Then" Clause has been actually executed.

---
In reply to: [357563259](https://github.com/microsoft/qsharp-runtime/pull/82#discussion_r357563259) [](ancestors = 357563259)